src/wheredataframe/wheredataframe.jl

```diff
@@ -0,0 +1,287 @@
+##
+## Show: show rows of parent that satisfies the `where` condition
+## with original row number
+##
+##############################################################################
+function Base.summary(wdf::WhereDataFrame)
+    @sprintf("%d×%d %s", size(parent(wdf))..., "WhereDataFrame")
+end
+
+Base.summary(io::IO, wdf::WhereDataFrame) = summary(wdf)
+
+
+function Base.show(io::IO,
+          wdf::WhereDataFrame;
+          allrows::Bool = !get(io, :limit, false),
+          allcols::Bool = !get(io, :limit, false),
+          splitcols = get(io, :limit, false),
+          rowlabel::Symbol = :Row,
+          eltypes::Bool = true,
+          truncate::Int = 32)
+    summary(io, wdf)
+    _show(io, wdf, allrows=allrows, allcols=allcols, splitcols=splitcols,
+          rowlabel=rowlabel, eltypes=eltypes, truncstring=truncate)
+end
+
+function _show(io::IO,
+               wdf::WhereDataFrame;
+               allrows::Bool = !get(io, :limit, false),
+               allcols::Bool = !get(io, :limit, false),
+               splitcols = get(io, :limit, false),
+               rowlabel::Symbol = :Row,
+               eltypes::Bool = true,
+               truncstring::Int)
+
+    df = parent(wdf)
+    _check_consistency(df)
+
+    # we will pass around this buffer to avoid its reallocation in ourstrwidth
+    buffer = IOBuffer(Vector{UInt8}(undef, 80), read=true, write=true)
+
+    nrows = length(rows(wdf))
+
+    dsize = displaysize(io)
+    availableheight = dsize[1] - 7
+    nrowssubset = fld(availableheight, 2)
+    bound = min(nrowssubset - 1, nrows)
+    if allrows || nrows <= availableheight
+        rowindices1 = rows(wdf)[1:nrows]
+        rowindices2 = 1:0
+    else
+        rowindices1 = rows(wdf)[1:bound]
+        rowindices2 = rows[max(bound + 1, nrows - nrowssubset + 1):nrows]
+    end
+    maxwidths = getmaxwidths(df, io, rowindices1, rowindices2, rowlabel, nothing,
+                             eltypes, buffer, truncstring)
+    width = getprintedwidth(maxwidths)
+    showrows(io, df, rowindices1, rowindices2, maxwidths, splitcols, allcols,
+             rowlabel, false, eltypes, nothing, buffer, truncstring)
+    return
+end
+
+
+##############################################################################
+##
+## Operations
+##
+##############################################################################
+
+Base.filter(wdf::WhereDataFrame) = parent(wdf)[rows(wdf), :]
+Base.filter!(wdf::WhereDataFrame) = delete!(parent(wdf), setdiff(1:nrow(parent(wdf)), rows(wdf)))
+Base.delete!(wdf::WhereDataFrame) = delete!(parent(wdf), rows(wdf))
+Base.view(wdf::WhereDataFrame) = view(parent(wdf), rows(wdf), :)
+combine(wdf::WhereDataFrame, args...; kwargs...) = combine(view(wdf), args...; kwargs...)
+DataFrame(wdf::WhereDataFrame; copycols::Bool=true) = DataFrame(view(wdf); copycols = copycols)
+DataAPI.describe(wdf::WhereDataFrame, args...; kwargs...) = describe(view(wdf), args...; kwargs...)
```

Comment:
since `where` is a wrapper (so `WhereDataFrame` is not an `AbstractDataFrame`) I think I would be better to avoid these methods and require `view` to be called by the user explicitly. Not sure - but this would better highlight that `where` is a lightweight wrapper only.

The point is that e.g. `transform(wdf)` and `combine(wdf, :)` will return a completely different result with your proposed design and I think it would be better to avoid this. Intuitively I would support only these methods for `WhereDataFrame` where its behavior differs from `SubDataFrame` and for other methods require to call `view` on it. Then the message to the user would be consistent.